Cannot build master with either Intel18 or Intel19 (on stampede2)

The stampede2 supercomputer provides the classic Intel compilers `16.0.3`, `17.0.4`, `18.0.0`, `18.0.2`, `19.1.1`.

As part of the release process, I try to compile the library with both `18.0.2` and `19.1.1` in a basic configuration with the corresponding IntelMPI package. This time I can't build deal.II on master with either of these. I tried to build [0516371](https://github.com/dealii/dealii/tree/051637127b117c3c50b215a6ab7a6183951dabfe).

---

Intel 18.0.2 (fixed by #15390)
-----

I get an error that `invalid_dof_index` does not exist, although we include `deal.II/base/types.h` in `dof_handler_policy.cc`. It is worth mentioning that this happens in a lambda function, and we used to have problems with Intel18 and lambda functions in the past.
```
/dealii/source/dofs/dof_handler_policy.cc(3533): error: namespace "dealii::numbers" has no member "invalid_dof_index"
                  if (*received_index != numbers::invalid_dof_index)
                                                  ^
          detected during:
            instantiation of function "lambda [](const auto &, const auto &)->auto [with <auto-1>=dealii::TriaActiveIterator<dealii::DoFCellAccessor<1, 1, false>>, <auto-2>=std::vector<dealii::types::global_dof_index={unsigned int}, std::allocator<dealii::types::global_dof_index={unsigned int}>>]" at line 2492 of "/opt/apps/gcc/6.3.0/include/c++/6.3.0/type_traits"
            instantiation of class "std::__result_of_impl<false, false, _Functor, _ArgTypes...> [with _Functor=lambda [](const auto &, const auto &)->auto &, _ArgTypes=<const dealii::TriaActiveIterator<dealii::DoFCellAccessor<1, 1, false>> &, const std::vector<dealii::types::global_dof_index={unsigned int}, std::allocator<dealii::types::global_dof_index={unsigned int}>> &>]" at line 2505 of "/opt/apps/gcc/6.3.0/include/c++/6.3.0/type_traits"
            instantiation of class "std::result_of<_Functor (_ArgTypes...)> [with _Functor=lambda [](const auto &, const auto &)->auto &, _ArgTypes=<const dealii::TriaActiveIterator<dealii::DoFCellAccessor<1, 1, false>> &, const std::vector<dealii::types::global_dof_index={unsigned int}, std::allocator<dealii::types::global_dof_index={unsigned int}>> &>]" at line 3563
            instantiation of "void dealii::internal::DoFHandlerImplementation::Policy::<unnamed>::communicate_dof_indices_on_marked_cells(const dealii::DoFHandler<dim, spacedim> &, std::vector<bool, std::allocator<bool>> &) [with dim=1, spacedim=1]" at line 3734
            instantiation of "dealii::internal::DoFHandlerImplementation::NumberCache dealii::internal::DoFHandlerImplementation::Policy::ParallelDistributed<dim, spacedim>::distribute_dofs() const [with dim=1, spacedim=1]" at line 24 of "/build-dealii-intel18-lld/source/dofs/dof_handler_policy.inst"

/dealii/source/dofs/dof_handler_policy.cc(3533): error: name followed by "::" must be a class or namespace name
                  if (*received_index != numbers::invalid_dof_index)
                                         ^
          detected during:
            instantiation of function "lambda [](auto &, auto)->auto [with <auto-1>=dealii::types::global_dof_index={unsigned int}, <auto-2>=dealii::types::global_dof_index={unsigned int} *]" at line 528 of "/dealii/include/deal.II/dofs/dof_accessor.templates.h"
            instantiation of "void dealii::internal::DoFAccessorImplementation::Implementation::process_object(const dealii::DoFHandler<dim, spacedim> &, unsigned int, unsigned int, dealii::types::fe_index={unsigned short}, const DoFMapping &, const std::integral_constant<int, structdim> &, dealii::types::global_dof_index={unsigned int} *&, const DoFProcessor &) [with dim=1, spacedim=1, structdim=0, DoFProcessor=lambda [](auto &, auto)->auto, DoFMapping=lambda [](auto)->auto]" at line 1183 of
                      "/dealii/include/deal.II/dofs/dof_accessor.templates.h"
            instantiation of "void dealii::internal::DoFAccessorImplementation::Implementation::DoFIndexProcessor<dim, spacedim>::process_vertex_dofs(dealii::DoFHandler<dim, spacedim> &, unsigned int, dealii::types::fe_index={unsigned short}, dealii::types::global_dof_index={unsigned int} *&, const DoFProcessor &) const [with dim=1, spacedim=1, DoFProcessor=lambda [](auto &, auto)->auto]" at line 1031 of "/dealii/include/deal.II/dofs/dof_accessor.templates.h"
            instantiation of "void dealii::internal::DoFAccessorImplementation::Implementation::process_dof_indices(const dealii::DoFAccessor<structdim, dim, spacedim, level_dof_access> &, const DoFIndicesType &, dealii::types::fe_index={unsigned short}, const DoFOperation &, const DoFProcessor &, bool) [with dim=1, spacedim=1, level_dof_access=false, structdim=1, DoFIndicesType=std::vector<dealii::types::global_dof_index={unsigned int}, std::allocator<dealii::types::global_dof_index={unsigned
                      int}>>, DoFOperation=dealii::internal::DoFAccessorImplementation::Implementation::DoFIndexProcessor<1, 1>, DoFProcessor=lambda [](auto &, auto)->auto]" at line 3563
            instantiation of "void dealii::internal::DoFHandlerImplementation::Policy::<unnamed>::communicate_dof_indices_on_marked_cells(const dealii::DoFHandler<dim, spacedim> &, std::vector<bool, std::allocator<bool>> &) [with dim=1, spacedim=1]" at line 3734
            instantiation of "dealii::internal::DoFHandlerImplementation::NumberCache dealii::internal::DoFHandlerImplementation::Policy::ParallelDistributed<dim, spacedim>::distribute_dofs() const [with dim=1, spacedim=1]" at line 24 of "/build-dealii-intel18-lld/source/dofs/dof_handler_policy.inst"


```

---

Intel 19.1.1 (fixed by #15425)
-----

There is some problem with the bundled TBB. Could it be that our updated cmake configuration broke something?
```
/dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp(83): error: class "tbb::spin_rw_mutex_v3::scoped_lock" has no member "internal_set_mutex"
              s.my_scoped_lock.internal_set_mutex(NULL);
                               ^

/dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp(129): error: class "tbb::spin_rw_mutex_v3::scoped_lock" has no member "internal_set_mutex"
                  s.my_scoped_lock.internal_set_mutex(this);  // need mutex for release()
                                   ^

/dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp(172): error: class "tbb::spin_rw_mutex_v3::scoped_lock" has no member "internal_set_mutex"
                  s.my_scoped_lock.internal_set_mutex(this);  // need mutex for release()
                                   ^

compilation aborted for /dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp (code 2)
```

**bad**: e4f621c35057603a83afc4dc843fd625b42a1d91 [detailed-bad.log](https://github.com/dealii/dealii/files/11827610/detailed-bad.log)

```
$ make VERBOSE=1 obj_tbb_debug
...
[100%] Building CXX object bundled/tbb-2018_U2/src/CMakeFiles/obj_tbb_debug.dir/tbb/x86_rtm_rw_mutex.cpp.o
cd /build-dealii/bundled/tbb-2018_U2/src && /cm/shared/apps/spack/cpu/opt/spack/linux-centos8-zen/gcc-8.3.1/intel-19.1.1.217-4d42ptjd6wsnh5bgbzcv6lp44vxpjwut/bin/icpc -DDEBUG -DDO_ITT_NOTIFY -DUSE_PTHREAD -D__TBB_BUILD=1 -I/build-dealii/bundled/tbb-2018_U2/src -I/dealii/bundled/tbb-2018_U2/src -I/dealii/bundled/tbb-2018_U2/src/rml/include -isystem /dealii/bundled/tbb-2018_U2/include -isystem /dealii/bundled/kokkos-3.7.00/algorithms/src -isystem /dealii/bundled/kokkos-3.7.00/containers/src -isystem /dealii/bundled/kokkos-3.7.00/core/src -isystem /dealii/bundled/kokkos-3.7.00/simd/src -isystem /dealii/bundled/kokkos-3.7.00/tpls/desul/include -isystem /dealii/bundled/muparser_v2_3_3/include -fpic -w2 -diag-disable=remark -diag-disable=16219 -wd21 -wd68 -wd135 -wd175 -wd177 -wd191 -wd193 -wd279 -wd327 -wd383 -wd854 -wd981 -wd1418 -wd1478 -wd1572 -wd2259 -wd2536 -wd2651 -wd3415 -wd15531 -wd15552 -wd111 -wd128 -wd185 -wd186 -wd280 -qopenmp-simd -pthread -Wno-parentheses -O0 -g -gdwarf-2 -grecord-gcc-switches -o CMakeFiles/obj_tbb_debug.dir/tbb/x86_rtm_rw_mutex.cpp.o -c /dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp
TBB Warning: tbb/task_scheduler_init.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.
TBB Warning: tbb/atomic.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.
/dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp(83): error: class "tbb::spin_rw_mutex_v3::scoped_lock" has no member "internal_set_mutex"
              s.my_scoped_lock.internal_set_mutex(NULL);
                               ^

/dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp(129): error: class "tbb::spin_rw_mutex_v3::scoped_lock" has no member "internal_set_mutex"
                  s.my_scoped_lock.internal_set_mutex(this);  // need mutex for release()
                                   ^

/dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp(172): error: class "tbb::spin_rw_mutex_v3::scoped_lock" has no member "internal_set_mutex"
                  s.my_scoped_lock.internal_set_mutex(this);  // need mutex for release()
                                   ^

compilation aborted for /dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp (code 2)
make[3]: *** [bundled/tbb-2018_U2/src/CMakeFiles/obj_tbb_debug.dir/build.make:498: bundled/tbb-2018_U2/src/CMakeFiles/obj_tbb_debug.dir/tbb/x86_rtm_rw_mutex.cpp.o] Error 2
make[3]: Leaving directory '/build-dealii'
make[2]: *** [CMakeFiles/Makefile2:1847: bundled/tbb-2018_U2/src/CMakeFiles/obj_tbb_debug.dir/all] Error 2
make[2]: Leaving directory '/build-dealii'
make[1]: *** [CMakeFiles/Makefile2:1854: bundled/tbb-2018_U2/src/CMakeFiles/obj_tbb_debug.dir/rule] Error 2
make[1]: Leaving directory '/build-dealii'
make: *** [Makefile:352: obj_tbb_debug] Error 2
```

---

**good**: 5f989f07bbd360f234f8eb75d564cf969cdf9fe8 [detailed-good.log](https://github.com/dealii/dealii/files/11827613/detailed-good.log)


```
$ make VERBOSE=1 obj_tbb_debug
...
[100%] Building CXX object bundled/tbb-2018_U2/src/CMakeFiles/obj_tbb_debug.dir/tbb/x86_rtm_rw_mutex.cpp.o
cd /build-dealii/bundled/tbb-2018_U2/src && /cm/shared/apps/spack/cpu/opt/spack/linux-centos8-zen/gcc-8.3.1/intel-19.1.1.217-4d42ptjd6wsnh5bgbzcv6lp44vxpjwut/bin/icpc -DDEBUG -DDO_ITT_NOTIFY -DUSE_PTHREAD -D__TBB_BUILD=1 -I/build-dealii/bundled/tbb-2018_U2/src -I/dealii/bundled/tbb-2018_U2/src -I/dealii/bundled/tbb-2018_U2/include -I/dealii/bundled/kokkos-3.7.00/algorithms/src -I/dealii/bundled/kokkos-3.7.00/containers/src -I/dealii/bundled/kokkos-3.7.00/core/src -I/dealii/bundled/kokkos-3.7.00/simd/src -I/dealii/bundled/kokkos-3.7.00/tpls/desul/include -I/dealii/bundled/muparser_v2_3_3/include -I/dealii/bundled/tbb-2018_U2/src/rml/include -fpic -w2 -diag-disable=remark -diag-disable=16219 -wd21 -wd68 -wd135 -wd175 -wd177 -wd191 -wd193 -wd279 -wd327 -wd383 -wd854 -wd981 -wd1418 -wd1478 -wd1572 -wd2259 -wd2536 -wd2651 -wd3415 -wd15531 -wd15552 -wd111 -wd128 -wd185 -wd186 -wd280 -qopenmp-simd -pthread -Wno-parentheses -O0 -g -gdwarf-2 -grecord-gcc-switches -o CMakeFiles/obj_tbb_debug.dir/tbb/x86_rtm_rw_mutex.cpp.o -c /dealii/bundled/tbb-2018_U2/src/tbb/x86_rtm_rw_mutex.cpp
```

--- a/bundled/boost-1.70.0/libs/iostreams/src/CMakeLists.txt
+++ b/bundled/boost-1.70.0/libs/iostreams/src/CMakeLists.txt
@@ -23,6 +23,8 @@ set(src_boost_iostreams
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-c11-extensions)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-deprecated-copy)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-uninitialized)
+enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-deprecated-builtins)
+enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-deprecated-declarations)
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../../../include)
 define_object_library(bundled_boost_iostreams OBJECT ${src_boost_iostreams})

--- a/bundled/boost-1.70.0/libs/iostreams/src/CMakeLists.txt
+++ b/bundled/boost-1.70.0/libs/iostreams/src/CMakeLists.txt
@@ -24,4 +24,5 @@ enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-c11-extensions)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-deprecated-copy)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-uninitialized)
 
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../../../include)
 define_object_library(bundled_boost_iostreams OBJECT ${src_boost_iostreams})

--- a/bundled/boost-1.70.0/libs/serialization/src/CMakeLists.txt
+++ b/bundled/boost-1.70.0/libs/serialization/src/CMakeLists.txt
@@ -57,6 +57,8 @@ set(src_boost_serialization
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-c11-extensions)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-deprecated-copy)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-uninitialized)
+enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-deprecated-builtins)
+enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-deprecated-declarations)
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../../../include)
 define_object_library(bundled_boost_serialization OBJECT ${src_boost_serialization})

--- a/bundled/boost-1.70.0/libs/serialization/src/CMakeLists.txt
+++ b/bundled/boost-1.70.0/libs/serialization/src/CMakeLists.txt
@@ -58,4 +58,5 @@ enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-c11-extensions)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-deprecated-copy)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-uninitialized)
 
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../../../include)
 define_object_library(bundled_boost_serialization OBJECT ${src_boost_serialization})

--- a/bundled/boost-1.70.0/libs/system/src/CMakeLists.txt
+++ b/bundled/boost-1.70.0/libs/system/src/CMakeLists.txt
@@ -18,5 +18,6 @@ set(src_boost_system
   error_code.cpp
   )
 
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../../../include)
 define_object_library(bundled_boost_system OBJECT ${src_boost_system})
 

--- a/bundled/kokkos-3.7.00/CMakeLists.txt
+++ b/bundled/kokkos-3.7.00/CMakeLists.txt
@@ -69,4 +69,12 @@ enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-missing-field-initializers)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-suggest-override)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-unused-but-set-parameter)
 
+include_directories(
+  ${CMAKE_CURRENT_SOURCE_DIR}/algorithms/src
+  ${CMAKE_CURRENT_SOURCE_DIR}/containers/src
+  ${CMAKE_CURRENT_SOURCE_DIR}/core/src
+  ${CMAKE_CURRENT_SOURCE_DIR}/simd/src
+  ${CMAKE_CURRENT_SOURCE_DIR}/tpls/desul/include
+  )
+
 define_object_library(bundled_kokkos OBJECT ${src_kokkos})

--- a/bundled/muparser_v2_3_3/CMakeLists.txt
+++ b/bundled/muparser_v2_3_3/CMakeLists.txt
@@ -24,9 +24,7 @@ enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-implicit-fallthrough)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-cast-function-type)
 enable_if_supported(DEAL_II_WARNING_FLAGS -Wno-float-conversion)
 
-include_directories(
-  include/
-)
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include/)
 
 define_object_library(bundled_muparser OBJECT
   src/muParserBase.cpp

--- a/bundled/tbb-2018_U2/src/CMakeLists.txt
+++ b/bundled/tbb-2018_U2/src/CMakeLists.txt
@@ -45,6 +45,7 @@ enable_if_supported(DEAL_II_WARNING_FLAGS "-Wno-tautological-overlap-compare")
 
 set(CMAKE_INCLUDE_CURRENT_DIR TRUE)
 include_directories(
+  ${CMAKE_CURRENT_SOURCE_DIR}/../include/
   ${CMAKE_CURRENT_SOURCE_DIR}/rml/include
   )
 

--- a/bundled/umfpack/AMD/Source/CMakeLists.txt
+++ b/bundled/umfpack/AMD/Source/CMakeLists.txt
@@ -30,6 +30,7 @@ set(src_amd
   amd_valid.cc
   )
 
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../Include)
 define_object_library(bundled_amd_int OBJECT ${src_amd})
 deal_ii_add_definitions(bundled_amd_int "DINT")
 

--- a/bundled/umfpack/UMFPACK/Source/CMakeLists.txt
+++ b/bundled/umfpack/UMFPACK/Source/CMakeLists.txt
@@ -348,6 +348,10 @@ set(src_umfpack_GENERIC
   umfpack_timer.cc
   )
 
+include_directories(
+  ${CMAKE_CURRENT_SOURCE_DIR}/../Include
+  ${CMAKE_CURRENT_SOURCE_DIR}/../../AMD/Include
+  )
 
 #-------------------------------------------------------------------------------
 # Compile each of the following files as long int routines; UMFPACK also allows

--- a/cmake/macros/macro_populate_target_properties.cmake
+++ b/cmake/macros/macro_populate_target_properties.cmake
@@ -62,9 +62,17 @@ function(populate_target_properties _target _build)
     ${CMAKE_BINARY_DIR}/include
     ${CMAKE_SOURCE_DIR}/include
     )
-  target_include_directories(${_target} SYSTEM PRIVATE
-    ${DEAL_II_BUNDLED_INCLUDE_DIRS}
-    )
+
+  #
+  # Do not add bundled include directories to bundled_ targets. First of
+  # all this is unnecessary, secondly, this severly trips up ICC-19 that
+  # cannot handle the additional -isystem include properly...
+  #
+  if(NOT "${_target}" MATCHES "^bundled_")
+    target_include_directories(${_target} SYSTEM PRIVATE
+      ${DEAL_II_BUNDLED_INCLUDE_DIRS}
+      )
+  endif()
 
   # Interface includes:
 

